70_FLIPFluids/main.cpp

```diff
@@ -0,0 +1,2240 @@
+				.levelCount = 1u,
+				.baseArrayLayer = 0u,
+				.layerCount = 1u
+			};
+			imageBarriers[0].oldLayout = IImage::LAYOUT::READ_ONLY_OPTIMAL;
+			imageBarriers[0].newLayout = IImage::LAYOUT::GENERAL;
+			imageBarriers[1].barrier = {
+				.dep = {
+					.srcStageMask = PIPELINE_STAGE_FLAGS::COMPUTE_SHADER_BIT,
+					.srcAccessMask = ACCESS_FLAGS::SHADER_WRITE_BITS | ACCESS_FLAGS::SHADER_READ_BITS,
+					.dstStageMask = PIPELINE_STAGE_FLAGS::COMPUTE_SHADER_BIT,
+					.dstAccessMask = ACCESS_FLAGS::SHADER_WRITE_BITS
+			}
+			};
+			imageBarriers[1].image = prevVelocityFieldImageView->getCreationParameters().image.get();
+			imageBarriers[1].subresourceRange = {
+				.aspectMask = IImage::EAF_COLOR_BIT,
+				.baseMipLevel = 0u,
+				.levelCount = 1u,
+				.baseArrayLayer = 0u,
+				.layerCount = 1u
+			};
+			imageBarriers[1].oldLayout = IImage::LAYOUT::READ_ONLY_OPTIMAL;
+			imageBarriers[1].newLayout = IImage::LAYOUT::GENERAL;
+
+			cmdbuf->pipelineBarrier(E_DEPENDENCY_FLAGS::EDF_NONE, {.imgBarriers = imageBarriers});
+		}
+
+		{
+			SMemoryBarrier memBarrier;
+			memBarrier.srcStageMask = PIPELINE_STAGE_FLAGS::ALL_TRANSFER_BITS;
+			memBarrier.srcAccessMask = ACCESS_FLAGS::MEMORY_WRITE_BITS;
+			memBarrier.dstStageMask = PIPELINE_STAGE_FLAGS::COMPUTE_SHADER_BIT;
+			memBarrier.dstAccessMask = ACCESS_FLAGS::SHADER_READ_BITS;
+			cmdbuf->pipelineBarrier(E_DEPENDENCY_FLAGS::EDF_NONE, {.memBarriers = {&memBarrier, 1}});
+        }
+
+		// simulation steps
+		for (uint32_t i = 0; i < m_substepsPerFrame; i++)
+		{
+			dispatchUpdateFluidCells(cmdbuf);			// particle to grid
+			dispatchApplyBodyForces(cmdbuf, i == 0);	// external forces, e.g. gravity
+			dispatchApplyDiffusion(cmdbuf);
+			dispatchApplyPressure(cmdbuf);
+			dispatchExtrapolateVelocities(cmdbuf);	// grid -> particle vel
+			dispatchAdvection(cmdbuf);				// update/advect fluid
+		}
+
+		{
+			SMemoryBarrier memBarrier;
+			memBarrier.srcStageMask = PIPELINE_STAGE_FLAGS::COMPUTE_SHADER_BIT;
+			memBarrier.srcAccessMask = ACCESS_FLAGS::SHADER_WRITE_BITS;
+			memBarrier.dstStageMask = PIPELINE_STAGE_FLAGS::COMPUTE_SHADER_BIT;
+			memBarrier.dstAccessMask = ACCESS_FLAGS::SHADER_READ_BITS;
+			cmdbuf->pipelineBarrier(E_DEPENDENCY_FLAGS::EDF_NONE, {.memBarriers = {&memBarrier, 1}});
+        }
+
+		// prepare particle vertices for render
+		cmdbuf->bindComputePipeline(m_genParticleVerticesPipeline.get());
+		cmdbuf->bindDescriptorSets(nbl::asset::EPBP_COMPUTE, m_genParticleVerticesPipeline->getLayout(), 1, 1, &m_genVerticesDs.get());
+		cmdbuf->dispatch(WorkgroupCountParticles, 1, 1);
+
+		{
+			SMemoryBarrier memBarrier;
+			memBarrier.srcStageMask = PIPELINE_STAGE_FLAGS::COMPUTE_SHADER_BIT;
+			memBarrier.srcAccessMask = ACCESS_FLAGS::SHADER_WRITE_BITS;
+			memBarrier.dstStageMask = PIPELINE_STAGE_FLAGS::VERTEX_INPUT_BITS;
+			memBarrier.dstAccessMask = ACCESS_FLAGS::MEMORY_READ_BITS;
+			cmdbuf->pipelineBarrier(E_DEPENDENCY_FLAGS::EDF_NONE, {.memBarriers = {&memBarrier, 1}});
+        }
```

Comment:
as stated before, the vertex shader could do all the quad generating work if you used an indexed draw

Reply:
also the pipeline barrier after the last compute dispatch and before renderpass begin shouldn't be here at all, should be handled via subpass FROM EXTERNAL dependency
https://github.com/Devsh-Graphics-Programming/Nabla-Examples-and-Tests/pull/145/files#r1797356123

Reply:
if you get rid of the `m_genParticleVerticesPipeline` dispatch then both barriers (line 1225 and 1211 should disappear)